tensorpipe/common/cuda.h

```diff
@@ -125,14 +127,66 @@ inline int cudaDeviceForPointer(const void* ptr) {
   return attrs.device;
 }
 
-using CudaPinnedBuffer = std::shared_ptr<uint8_t>;
+class CudaPinnedBufferAllocator {
+ public:
+  CudaPinnedBufferAllocator(const CudaPinnedBufferAllocator&) = delete;
+  CudaPinnedBufferAllocator(CudaPinnedBufferAllocator&&) = delete;
+  CudaPinnedBufferAllocator& operator=(const CudaPinnedBufferAllocator&) =
+      delete;
+  CudaPinnedBufferAllocator& operator=(CudaPinnedBufferAllocator&&) = delete;
+
+  CudaPinnedBufferAllocator(size_t numChunks = 1024, size_t chunkSize = 1024)
+      : kNumChunks_(numChunks),
+        kChunkSize_(chunkSize),
+        chunkAvailable_(numChunks, false),
+        basePtr_(allocPinnedBuffer(numChunks * chunkSize)) {}
+
+  std::shared_ptr<uint8_t> getBuffer(size_t size) {
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      size_t requestedChunks = (size + kChunkSize_ - 1) / kChunkSize_;
+      size_t startChunk = 0;
+
+      // Linear check for available contiguous chunks.
+      for (size_t curChunk = 0; curChunk < kNumChunks_; ++curChunk) {
+        if (!chunkAvailable_[curChunk]) {
+          startChunk = curChunk + 1;
+        } else if (curChunk - startChunk + 1 == requestedChunks) {
+          for (size_t i = startChunk; i <= curChunk; ++i) {
+            chunkAvailable_[i] = false;
+          }
+
+          return std::shared_ptr<uint8_t>(
+              basePtr_.get() + startChunk * kChunkSize_,
+              [this, startChunk, curChunk](uint8_t* /* unused */) {
+                for (size_t i = startChunk; i <= curChunk; ++i) {
+                  // There is a harmless race condition here.
+                  chunkAvailable_[i] = true;
+                }
+              });
+        }
+      }
+    }
 
-inline CudaPinnedBuffer makeCudaPinnedBuffer(size_t length) {
-  void* ptr;
-  TP_CUDA_CHECK(cudaMallocHost(&ptr, length));
-  return CudaPinnedBuffer(reinterpret_cast<uint8_t*>(ptr), [](uint8_t* ptr) {
-    TP_CUDA_CHECK(cudaFreeHost(ptr));
-  });
-}
+    // No contiguous pre-allocated chunks available, falling back to allocating
+    // a temporary buffer.
+    return allocPinnedBuffer(size);
```

Comment:
I'm worried that this may be problematic, as it means that the original problem we're trying to solve could still occur, less frequently but more unpredictably.

You said that one tradeoff you considered is between the cost of one cudaHostMalloc call and multiple cudaMemcpyAsync calls, and I guess you deemed the latter to be more undesirable than the former. I however am leaning the other way, as a cudaHostMalloc delays _all_ the kernels that the process is launching, even those that are completely unrelated to TP. In other words, it completely breaks any computation/communication overlap the user might be trying to do.

On the other hand, I think we can mitigate and amortize the cost of cudaMemcpyAsync by having a large enough buffer and "pipelining" correctly: since I expect the bottleneck to be TCP (since PCIe should hopefully be faster), we should have some slack in our PCIe transfers and thus we should be able to "hide" some overhead.

I also believe that we shouldn't worry too much about delaying some transfers when the buffer is full (i.e., having those channels wait for a slot to free up, rather than allocating new memory). That's because if the buffer is full it means that there already are many transfers happening, and thus even if we immediately added an additional one it may not receive bandwidth until the previous ones finish.

This clearly requires a rework of the channel logic, in order for it to request a buffer and deal with the possibility of it not being available right away, and thus giving a callback in order to be woken up later.